src/lazy/text/buffer.rs

```diff
@@ -107,37 +109,52 @@ fn dbg_parse<I: Debug, O: Debug, E: Debug, P: Parser<I, O, E>>(
 /// pair. The `match` may be either the slice of the input that was matched (represented as another
 /// `TextBufferView`) or a `MatchedValue` that retains information discovered during parsing that
 /// will be useful if the match is later fully materialized into a value.
-#[derive(PartialEq, Clone, Copy)]
-pub(crate) struct TextBufferView<'a> {
+#[derive(Clone, Copy)]
+pub(crate) struct TextBufferView<'top> {
     // `data` is a slice of remaining data in the larger input stream.
     // `offset` is the absolute position in the overall input stream where that slice begins.
     //
     // input: 00 01 02 03 04 05 06 07 08 09
     //                          └────┬────┘
     //                          data: &[u8]
     //                          offset: 6
-    data: &'a [u8],
+    data: &'top [u8],
     offset: usize,
+    allocator: &'top BumpAllocator,
```

Comment:
🗺️ `TextBufferView` now holds a reference to the bump allocator, allowing it to cheaply store structural information about containers and e-expressions. Data living in the bump allocator has the `'top` lifetime.

Passing an allocator into each parsing method as an argument would cause those methods to violate the `Parser` contract, which says they must only accept a single parameter of some type `I: Input`. Adding that field allows us to use it everywhere that a `TBV` is already accepted.